File: Python/Permutation-Sequence.py
class Solution:
    # @return a string
    def factorial(self, num):
        if num <= 1:
            return 1
        cnt = 1
        for i in range(1,num+1):
            cnt *= i
        return cnt

    def get(self, vis, pos):
        cnt = 0
        for i in range(len(vis)):
            if pos == cnt and vis[i] == 0:
                return i + 1
            if vis[i] == 0:
                cnt += 1

    def getPermutation(self, n, k):
        vis = [0] * n
        s = ""
        num = 0
        while num < n:
            cnt = self.factorial(n - num - 1)
            res = k / cnt
            if k / cnt != 0 and k % cnt == 0:
                res -= 1
            k -= res * cnt
            ans = self.get(vis, res)
            s += str(ans)
            vis[ans-1] = 1
            num += 1
        return s

File: C++/Permutation-Sequence.cpp
#include <string>
#include <vector>

class Solution {
public:
    /**
      * @param n: n
      * @param k: the kth permutation
      * @return: return the k-th permutation
      */
    int factor(int x) {
        int res = 1;
        for (int i = 2; i <= x; i ++) {
            res *= i;
        }
        return res;
    }
    int get(vector<int> &vis, int pos) {
        int cnt = 0;
        for (int i = 0; i < vis.size(); i ++) {
            if (pos == cnt && vis[i] == 0) return i + 1;
            else if (vis[i] == 0) cnt ++;
        }
    }
    string getPermutation(int n, int k) {
        vector<int> vis(n);
        string s;
        int num = 0;
        while (num < n) {
            int cnt = factor(n - num - 1);
            int res = k / cnt;
            if (res != 0 && k % cnt == 0) {
                res --;
            }
            k -= res * cnt;
            int ans = get(vis, res);
            char ch = '0' + ans;
            s += ch;
            vis[ans-1] = 1;
            num += 1;
        }
        return s;
    }
};
